### PR TITLE
[Kernel] CUDA CSR2COO COOSort COO2CSR

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

* Add CSR2COO, COOSort and COO2CSR using cusparse.
* Add several array operators needed in CUDA.
* Change the ATEN_XPU_SWITCH to have a more informative error message when the device is not supported.
* Add operator overloading for DLDataType and DLContext.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by jermainewang** [comment]
@VoVAllen  please help me test compilation on macos.

**Comment by VoVAllen** [comment]
It works on my macos.

File: include/dgl/runtime/ndarray.h [hunk 1]
```diff
@@ -210,6 +203,12 @@ class NDArray {
    * \brief Get the size of the array in the number of bytes.
    */
   size_t GetSize() const;
+
+  /*!
+   * \brief Get the number of elements in this array.
+   */
+  int64_t Numel() const;
```
Comment (classicsong) [inline include/dgl/runtime/ndarray.h:210]: NumEl ?

File: tests/cpp/test_spmat.cc [hunk 2]
```diff
@@ -252,15 +252,15 @@ TEST(SpmatTest, TestCSRTranspose) {
 }
 
 template <typename IDX>
-void _TestCSRToCOO() {
-  auto csr = CSR2<IDX>();
+void _TestCSRToCOO(DLContext ctx) {
```
Comment (classicsong) [inline tests/cpp/test_spmat.cc:255]: Do we miss the GPU test for _TestCSRTranspose?

**Pull request merged**